### PR TITLE
Add `rb_hash_new`, `rb_hash_aset`, and `rb_hash_foreach`

### PR DESCRIPTION
**Pull request by wagenet**
Are any of these a macro? NO

Is there a macro version for any of these? NO

Defined in:

2.3 https://github.com/ruby/ruby/blob/v2_3_7/include/ruby/intern.h#L512
2.4 https://github.com/ruby/ruby/blob/v2_4_4/include/ruby/intern.h#L509
2.5 https://github.com/ruby/ruby/blob/v2_5_1/include/ruby/intern.h#L491
2.6 https://github.com/ruby/ruby/blob/v2_6_0_preview2/include/ruby/intern.h#L491

### CODE REVIEW

File: src/intern.rs [hunk 1]
```diff
@@ -6,6 +6,10 @@ extern {
     pub fn rb_ary_new_capa(capacity: c_long) -> VALUE;
     pub fn rb_ary_push(array: VALUE, item: VALUE) -> VALUE;
 
+    pub fn rb_hash_new() -> VALUE;
+    pub fn rb_hash_aset(hash: VALUE, key: VALUE, val: VALUE) -> VALUE;
+    pub fn rb_hash_foreach(hash: VALUE, func: extern "C" fn(key: VALUE, val: VALUE, farg: VALUE), farg: VALUE);
```
Comment (chancancode) [inline src/intern.rs]: This is not quite right. At minimum `func` should be typed as returning a `c_int`, although the header actually typed it as an `ANYARGS<c_int>`. I agree that we _probably_ want to be more precise here and match the actual `rb_foreach_func` definition (should mention that in the notes section of the documentation).

Comment (chancancode) [reply src/intern.rs]: UB: when the func returns anything other than the `ST_*` values

**Comment by wagenet** [comment]
Should note that it will raise `RuntimeError` if the iterator manages to modify the hash.

File: src/intern.rs (hunk 1 above)
Comment (wagenet) [reply src/intern.rs]: Yeah, I copied some of this from Helix which I think does it wrong. For reference:
```c
typedef int rb_foreach_func(VALUE, VALUE, VALUE);
```

As for `ST_*` should we just make our own version of the enum?

File: ext/ruby.c [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #include "ruby/ruby.h"
 #include "stdint.h"
+#include "intern.c"
```
Comment (chancancode) [inline ext/ruby.c]: I don’t think we need this. The makefile seems to be able to find all c files just fine.

File: src/intern.rs [hunk 3]
```diff
@@ -1,7 +1,24 @@
 use super::*;
 use libc::{c_char, c_int, c_long};
 
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
```
Comment (chancancode) [inline src/intern.rs]: Probably don’t need Ord/PartialOrd/Hash for this one

File: ext/ruby.c (hunk 2 above)
Comment (wagenet) [reply ext/ruby.c]: Doesn't build without it for me.

Comment (wagenet) [reply ext/ruby.c]: Wait... if I add `#include "ruby/ruby.h"` to `intern.c` then it does... strange.

Comment (wagenet) [reply ext/ruby.c]: I feel like something strange is going on here.

Comment (wagenet) [reply ext/ruby.c]: Oh, I see, `ruby/intern.h` doesn't include `ruby/ruby.h` so I have to include for it.

**Pull request merged**